src/sync_backend_clients/gitlab_sync_backend_client.test.js

```diff
@@ -7,9 +7,11 @@ import {
 
 test('Parses GitLab project from URL', () => {
   [
-    ['https://gitlab.com/user/foo', 'user%2Ffoo'],
-    ['https://gitlab.com/group/subgroup/project', 'group%2Fsubgroup%2Fproject'],
-    ['gitlab.com/foo/bar', 'foo%2Fbar'],
+    ['https://gitlab.com/user/foo', ['gitlab.com', 'user%2Ffoo']],
+    ['https://gitlab.com/group/subgroup/project', ['gitlab.com', 'group%2Fsubgroup%2Fproject']],
+    ['https://gitlab.example.com/user/foo', ['gitlab.example.com', 'user%2Ffoo']],
+    ['gitlab.com/foo/bar', ['gitlab.com', 'foo%2Fbar']],
+    ['gitlab.example.com/foo/bar', ['gitlab.example.com', 'foo%2Fbar']],
     ['gitlab.com/user-but-no-project', undefined],
```

Comment:
Not your code @MaxSchlueter but I just saw this: Isn't it uncommon to use `undefined` deliberately @chasecaleb @munen ? I'm not a professional JS dev but I thought `undefined` is e.g. for a forgotten return statement or an undefined variable. Wouldn't `null` be the better option here? I mean, the parse function should return null if not successful?

Reply:
Hi @schoettl, no idea how I missed your ping all those months ago :facepalm:

>  Imo the author of the original function wanted to be explicit about it, since it's an error case. 

Yes, this is why I didn't just do `return` -- plus I generally work in languages that have a void type so they don't allow you to mix a plain `return` and `return <something>` in the same function. 

But as for specifically picking `undefined` instead of `null`... meh :shrug:. Around the time I wrote this code I was working in a TypeScript codebase that banned `null` and stuck with only `undefined` for unimportant reasons.

In my opinion deciding whether to use `null` vs `undefined` doesn't really matter beyond perhaps picking one for style/consistency reasons. Instead, it's better to write the calling code to handle both equivalently... for example, `if (foo == null) { ... }` is true if `foo` is either `null` or `undefined`. This goes doubly so for JavaScript, as opposed to Typescript where you at least _might_ know whether a value could be null vs undefined vs neither.

Here's an interesting read: https://basarat.gitbook.io/typescript/recap/null-undefined . You'll notice that Basarat agrees with you about using null, @schoettl... but at the end he also points out the TypeScript team only uses undefined and [Douglas Crockford](https://en.wikipedia.org/wiki/Douglas_Crockford) also says to avoid null.

The _real_ best practice is to use a language with neither `null` nor `undefined` :grin:  I'll see myself out now...

Reply:
Hi @chasecaleb 

Thanks for chiming in and giving some detailed feedback!

I like your point about writing code that handle both cases equivalently. That's very common in dynamic languages.

For sake of completeness, and future readers, I'll say that the linked [interesting read](https://basarat.gitbook.io/typescript/recap/null-undefined) doesn't completely apply to JS, as it is meant for a TS audience. For example "Don't use undefined as a means of denoting validity". For statically typed language, it's the right call. And it's reasonable to explicitly point it out when working from a dynamically typed language into a statically typed language (as is the case with TS). In a purely dynamically typed language (like JS, Clojure, Ruby, etc), it's quite common and good practice. If anyone asks why: They have different tradeoffs. Check out the example code of the article above. It's arguable that the first example, a simple one-liner is very easy to read and reason about. The second function has  a body of 7 lines, that's quite a big hit to readability. Having said so, it's the style of statically typed language, _because_ it yields a type that's easy to reason about. That's a pro, too. Just a different one. You might be inclined to like one more or the latter. Everything is ok :pray:

Reply:
Alright, so now fixed conventions about this in organice. I personally used `x === null` for my null checks because `==` generally feels so sloppy in JS and I always fear that beginners pick it up and use it for other types like `x == ''`…